## Introduction to Programming



Welcome to the wonderfull world of programming. Programming and knowing a programming language is one of the most useful, employable and transferable skills you can have. Many of you will already have some coding experience either doing statistics in R or maybe using Matlab. Here you will learn from the foundations some of the core coding skills that you will need for scientific computing and will give you a firm grounding in the fundementals of programming which can dramtically help with career advancement

Programming is all about logic, you provide logical instructions to the computer in the programming language and it gives you an output... hopefully the one you want.

A lot of programming uses words logic that we would find familiar if maybe a bit formal:

"Do this, then do that"

"IF this, do this"

A useful tool in programming is to write out what you want to do in english or half code half english and then put the correct code into it. This is called pseudo code.

We will be using a language called Python. Python is an interpreted high level programming language. Interpreted means that programs written in python are not compiled so can be run line by line, and high level means that it does not have similar syntax to the base programming code of the computer like c++. 

We will start with some basic pseudo code for what we want and see if it works. I want the computer to say hello to me.

In [7]:
say hello

SyntaxError: invalid syntax (<ipython-input-7-5cb20dbb486a>, line 1)

Well that isn't anything...

To do anything in python it needs to be a defined function or object meaning that it must be known in the python language. You can't just make up a word in english and expect people to understand it. In this case say is not a word that python knows.

So we will be using functions to perform tasks.

When it comes to programming to start with a lot of the hard work has been done for us in that it comes with a wide range of in built functions and other nice people have done lots of work building libraries with functions they have written.

We will start with the print function

In [3]:
print("Hello")

Hello


You said hello to the computer and it said it back... while in reality what you did was use a base python function print. This function just prints out into the terminal whatever you tell it to. It is probably the single most useful function for error checking your work. 

Functions in python are like functions in maths. You have an input the function is applied to and then the output is given to you. 

In python the input of a function has to be specified in brackets attached to that function so print(input) applies the print function of writing out into the terminal to whatever is inside the brackets.

In this case we got it to print the word Hello however if we tried it without the "" would it work?

In [8]:
print(Hello)

NameError: name 'Hello' is not defined

It says Hello is not defined. This means that it doesn't know what Hello is. It thinks it is an object or a function and cannot find it. When we put things in "" python knows that it is text and treats it as such so you can make it print anything in "".

In [11]:
print("anything")

anything


In [13]:
print("Whatever you want can go inside quotation marks")

Whatever you want can go inside quotation marks


In [15]:
print("beep boop I'm a computer")

beep boop I'm a computer


There are numerous functions in python that can be used but the point of programming is to combine the functions that exist into a program that you want. Like combining words into a sentence.

Python has a huge variety of functions within it and available through other people's hard work. 

Another fundimental part of python is expressions - most simply these are mathematical expressions where python comes with it's own inbuilt maths solver.

In [17]:
1 + 1

2

This is a simple mathematical expression - an expression is a segment of code which is made up of values (numbers) and operators (+) which are in themselves a form of function. Python then evaluates this expression in that it will eventually reduce down to just a single output. 

Expressions can look complex being made of many values and operators.

In [19]:
(5*17) + (12/45) - (32 ** 3)

-32682.733333333334

essentially python is iteratively solving things based on BODMAS 

(5*17) + (12/45) - (32 ** 3)

85 + 0.266666 - 32768

-32682.733333333334

There are a number of operators within python:

Addition +

Subtraction -

Division /

Multiplication *

Exponent **

You will learn more over time.



The next thing to start learning is the types of data. Python is very sensitive to the type of data that is being dealt with. Applying functions to a data type that it does not expect will give an error. 

There are numerous types within python we will cover some of the basic types here. 

We have already seen a couple types and we will go over what they were

"Hello"

This is plain text and as such is being treated as the type String - this just means that it is a series of characters.

In [25]:
type("Hello")

str

The function type is a very simple function which tells you the type that the input is.

Strings are a very useful type of object, they are used for when anything is plain text words, commonly being used right now for website scraping etc. You can extrac portions of strings or combine them to make sentences etc. They are generally used for outputs during work flow.


In [32]:
"Hello" + "World"

'HelloWorld'

See we can combine strings to make more strings. 

In [31]:
"Hello" * "World"

TypeError: can't multiply sequence by non-int of type 'str'

However this is the only opperator that we can apply to strings. 

The next type of data we will look at is the integer. As with maths integers are non decimal numbers. 

In [47]:
type(1)

int

int is short for type integer. This type is generally used for numbers but the important thing to remember is that if it is an integer it cannot contain decimals.

In [36]:
type(1.4)

float

When there are decimals within the numbers it becomes type float. Float is generally the most common numeric type you will be using as it allows decimals. 

The next type is a boolean - a boolean is a true or false value, booleans are generally used for checking if things are a certain way in an if statement etc.

In [40]:
type(True)

bool

The next type we will look at is the list. Lists are basically a vector of different values combined into one. 

In [46]:
type([1,2,3])

list

Lists are used when you need to store multiple variables in one place and need to be able to manipulate them individually. Often you will make lists out of other data you have been handling such as integers or floats. 

Try and see what you can do with data types. See if you can combine types make lists etc. just play around with the type() command putting different things into them.